harness/determined/pytorch/_pytorch_context.py

```diff
@@ -85,6 +89,16 @@ def wrap_model(self, model: torch.nn.Module) -> torch.nn.Module:
         model_id = len(self.models)
         self._main_model.__setattr__(f"model_{model_id}", model)
 
+        if self._auto_amp:
+            # Ensure the forward pass is wrapped in an autocast context:
+            model.forward = torch.cuda.amp.autocast()(model.forward)
+
+            # Eliminate any need for the loss functions to be in that context:
+            def hook(module, input, output):
+                return output.float32() if output.dtype == torch.float16 else output
```

Comment:
Actually this check if it's a Tensor, and only then type and change the dtype. BERT returns what is essentially a dictionary - not sure if we then need to worry about ITS internal data types?

Reply:
Oh snap.  Probably just make this recursive to handle the basic nesting types (dict, tuple list) like we do with `pytorch._data.to_device()`.  That'll cover nearly everyone.

Reply:
Yeah when I started doing my big test with this (BERT) I saw this need to handle things other than Tensors. BERT sends back a dictionary, but yeah I guess I need to check the types inside that too...